File: July2022/C++/ValidAnagram.cpp
class Solution {
public:
    // TC: O(n)
    // SC: O(26) -> O(1)
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }
        
        vector<int> letters(26, 0);
        
        char c1, c2;
        for (int i = 0; i < s.size(); i++) {
            c1 = s[i];
            c2 = t[i];
            
            letters[c1 - 'a']++;    // incrementing the char count  
            letters[c2 - 'a']--;    // decrementing the char count
        }
        
        // if they are anagram then the array should contain 0 value
        for (int i = 0; i < 26; i++) {
            if (letters[i] != 0) {
                return false;
            }
        }
        
        return true;
    }
};

File: July2022/Java/ValidAnagram.java

// @saorav21994
// TC : O(n)
// SC : O(26) ~ O(1)

class Solution {
    public boolean isAnagram(String s, String t) {
        int hash[] = new int[26];
        int sl = s.length();
        int tl = t.length();
        for (int i = 0; i < sl; i++) {
            hash[(int)s.charAt(i) - 97] += 1;
        }
        for (int i = 0; i < tl; i++) {
            hash[(int)t.charAt(i) - 97] -= 1;
        }
        for (int i = 0; i < 26; i++) {
            if (hash[i] != 0) return false;
        }
        return true;
    }
}
